src/corehost/cli/fxr/corehost_init.h

```diff
@@ -0,0 +1,173 @@
+        }
+
+        for (const auto& kv : combined_properties)
+        {
+            m_clr_keys.push_back(kv.first);
+            m_clr_values.push_back(kv.second);
+        }
+
+        make_cstr_arr(m_fx_names, &m_fx_names_cstr);
+        make_cstr_arr(m_fx_dirs, &m_fx_dirs_cstr);
+        make_cstr_arr(m_fx_requested_versions, &m_fx_requested_versions_cstr);
+        make_cstr_arr(m_fx_found_versions, &m_fx_found_versions_cstr);
+        make_cstr_arr(m_clr_keys, &m_clr_keys_cstr);
+        make_cstr_arr(m_clr_values, &m_clr_values_cstr);
+    }
+
+    const pal::string_t& tfm() const
+    {
+        return m_tfm;
+    }
+
+    const host_interface_t& get_host_init_data()
+    {
+        host_interface_t& hi = m_host_interface;
+
+        hi.version_lo = HOST_INTERFACE_LAYOUT_VERSION_LO;
+        hi.version_hi = HOST_INTERFACE_LAYOUT_VERSION_HI;
+
+        hi.config_keys.len = m_clr_keys_cstr.size();
+        hi.config_keys.arr = m_clr_keys_cstr.data();
+
+        hi.config_values.len = m_clr_values_cstr.size();
+        hi.config_values.arr = m_clr_values_cstr.data();
+
+        // Keep these for backwards compat
+        if (m_fx_names_cstr.size() > 1)
+        {
+            hi.fx_name = m_fx_names_cstr[1];
+            hi.fx_dir = m_fx_dirs_cstr[1];
+            hi.fx_ver = m_fx_requested_versions_cstr[1];
+        }
+        else
+        {
+            hi.fx_name = _X("");
+            hi.fx_dir = _X("");
+            hi.fx_ver = _X("");
+        }
+
+        hi.deps_file = m_deps_file.c_str();
+        hi.additional_deps_serialized = m_additional_deps_serialized.c_str();
+        hi.is_framework_dependent = m_is_framework_dependent;
+
+        hi.probe_paths.len = m_probe_paths_cstr.size();
+        hi.probe_paths.arr = m_probe_paths_cstr.data();
+
+        // These are not used anymore, but we have to keep them for backward compat reasons.
+        // Set default values.
+        hi.patch_roll_forward = true;
+        hi.prerelease_roll_forward = false;
+
+        hi.host_mode = m_host_mode;
+
+        hi.tfm = m_tfm.c_str();
+
+        hi.fx_names.len = m_fx_names_cstr.size();
+        hi.fx_names.arr = m_fx_names_cstr.data();
+
+        hi.fx_dirs.len = m_fx_dirs_cstr.size();
+        hi.fx_dirs.arr = m_fx_dirs_cstr.data();
+
+        hi.fx_requested_versions.len = m_fx_requested_versions_cstr.size();
+        hi.fx_requested_versions.arr = m_fx_requested_versions_cstr.data();
+
+        hi.fx_found_versions.len = m_fx_found_versions_cstr.size();
+        hi.fx_found_versions.arr = m_fx_found_versions_cstr.data();
+
+        hi.host_command = m_host_command.c_str();
+
+        hi.host_info_host_path = m_host_info_host_path.c_str();
+        hi.host_info_dotnet_root = m_host_info_dotnet_root.c_str();
+        hi.host_info_app_path = m_host_info_app_path.c_str();
+
+        return hi;
+    }
+
+private:
+
+    static void make_cstr_arr(const std::vector<pal::string_t>& arr, std::vector<const pal::char_t*>* out)
```

Comment:
I know this is a refactor, but can we please get away from private static constructs. They are completely unnecessary and break incremental build semantics when one is simply updating implementation in a compilation unit rather than mucking with the interface. Just move the `make_cstr_arr` function into the compilation unit and it will work.

Reply:
I'll move all the bodies into a cpp… in a different change to keep things organized.